File: Sorting/quick sort algorithm/quick_sort.js
// One of the fastest sorting algorithm exist
function quickSort(arr, low, high) {
  if (low < high) {
    const pivotIndex = partition(arr, low, high);
    quickSort(arr, low, pivotIndex);
    quickSort(arr, pivotIndex + 1, high);
  }
}

function partition(arr, low, high) {
  const pivot = arr[low];
  let left = low + 1;
  let right = high;

  while (true) {
    while (left <= right && arr[left] <= pivot) {
      left++;
    }
    while (arr[right] >= pivot && right >= left) {
      right--;
    }
    if (right < left) {
      break;
    } else {
      [arr[left], arr[right]] = [arr[right], arr[left]];
    }
  }

  [arr[low], arr[right]] = [arr[right], arr[low]];
  return right;
}

// Example usage
let array = [12, 11, 13, 5, 6, 7];
quickSort(array, 0, array.length - 1);
console.log("Sorted array:", array);

File: Sorting/quick sort algorithm/quick_sort.py
# One of the fastest sorting algorithm exist
def quick_sort(arr, low, high):
    if low < high:
        pivot_index = partition(arr, low, high)
        quick_sort(arr, low, pivot_index)
        quick_sort(arr, pivot_index + 1, high)


def partition(arr, low, high):
    pivot = arr[low]
    left = low + 1
    right = high

    done = False
    while not done:
        while left <= right and arr[left] <= pivot:
            left = left + 1
        while arr[right] >= pivot and right >= left:
            right = right - 1
        if right < left:
            done = True
        else:
            arr[left], arr[right] = arr[right], arr[left]

    arr[low], arr[right] = arr[right], arr[low]
    return right


# Example usage
array = [12, 11, 13, 5, 6, 7]
quick_sort(array, 0, len(array) - 1)
print("Sorted array:", array)
